p2p/trust/trustmetric.go

```diff
@@ -0,0 +1,545 @@
+// Copyright 2017 Tendermint. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package trust
+
+import (
+	"encoding/json"
+	"math"
+	"sync"
+	"time"
+
+	cmn "github.com/tendermint/tmlibs/common"
+	dbm "github.com/tendermint/tmlibs/db"
+)
+
+// TrustMetricStore - Manages all trust metrics for peers
+type TrustMetricStore struct {
+	cmn.BaseService
+
+	// Maps a Peer.Key to that peer's TrustMetric
+	peerMetrics map[string]*TrustMetric
+
+	// Mutex that protects the map and history data file
+	mtx sync.Mutex
+
+	// The db where peer trust metric history data will be stored
+	db dbm.DB
+
+	// This configuration will be used when creating new TrustMetrics
+	config TrustMetricConfig
+}
+
+// NewTrustMetricStore returns a store that saves data to the DB
+// and uses the config when creating new trust metrics
+func NewTrustMetricStore(db dbm.DB, tmc TrustMetricConfig) *TrustMetricStore {
+	tms := &TrustMetricStore{
+		peerMetrics: make(map[string]*TrustMetric),
+		db:          db,
+		config:      tmc,
+	}
+
+	tms.BaseService = *cmn.NewBaseService(nil, "TrustMetricStore", tms)
+	return tms
+}
+
+// OnStart implements Service
+func (tms *TrustMetricStore) OnStart() error {
+	tms.BaseService.OnStart()
+
+	tms.mtx.Lock()
+	defer tms.mtx.Unlock()
+
+	tms.loadFromDB()
+	return nil
+}
+
+// OnStop implements Service
+func (tms *TrustMetricStore) OnStop() {
+	tms.mtx.Lock()
+	defer tms.mtx.Unlock()
+
+	// Stop all trust metric goroutines
+	for _, tm := range tms.peerMetrics {
+		tm.Stop()
+	}
+
+	tms.saveToDB()
```

Comment:
Should we saveToDB more frequently so we don't lose data if there's a crash? Perhaps a separate go-routine that persists this info every minute or so?

Reply:
agreed. It will be in the next commit